pkg/util/grpc/timeout.go

```diff
@@ -0,0 +1,29 @@
+package grpc
+
+import (
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// DoWithTimeout runs f and returns its error.  If the deadline d elapses first,
+// it returns a grpc DeadlineExceeded error instead.
+func DoWithTimeout(f func() error, d time.Duration) error {
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- f()
+		close(errChan)
+	}()
+
+	t := time.NewTimer(d)
+	select {
+	case <-t.C:
+		return status.Errorf(codes.DeadlineExceeded, "timed out waiting for operation")
+	case err := <-errChan:
+		if !t.Stop() {
+			<-t.C
+		}
```

Comment:
Why do we wait for the time left in the timer ?

Reply:
Best practices according to the docs: https://golang.org/pkg/time/#Timer.Stop

Reply:
I missed the `Stop` above :facepalm: